sklearn/metric_learning/nca.py

```diff
@@ -0,0 +1,277 @@
+from ..base import BaseEstimator, TransformerMixin
+from ..utils.validation import check_is_fitted
+
+import numpy as np
+import scipy as sp
+import scipy.misc
+import scipy.optimize
+import time
+
+import sys
+
+
+def nca_vectorized_oracle(L, X, y, n_components, loss, outer, threshold=0.0):
+    n_samples, n_features = X.shape
+
+    L = L.reshape((n_components, n_features))
+
+    Lx = np.dot(X, L.T)  # n_samples x n_components
+    assert Lx.shape == (n_samples, n_components)
+
+    A = Lx[np.newaxis, :, :] - Lx[:, np.newaxis, :]  # n_samples x n_samples x n_components
+    assert A.shape == (n_samples, n_samples, n_components)
+
+    logp = -np.einsum("ijk,ijk->ij", A, A)  # n_samples x n_samples
+    np.fill_diagonal(logp, -np.inf)
+    logp -= sp.misc.logsumexp(logp, axis=1)[:, np.newaxis]
+    assert logp.shape == (n_samples, n_samples)
+
+    p = np.exp(logp)  # n_samples x n_samples
+    assert p.shape == (n_samples, n_samples)
+
+    class_neighbours = y[:, np.newaxis] == y[np.newaxis, :]
+
+    p_i = np.einsum("ij,ij->i", p, class_neighbours)
+
+    proba = np.copy(p)
+    if loss == 'l1':
+        proba *= p_i[:, np.newaxis]
+    mask = proba > threshold
+    proba *= mask
+
+    grad = np.einsum("ij,ijkl", proba, outer)
+    assert grad.shape == (n_features, n_features)
+
+    neighbours_proba = class_neighbours * p
+    if loss == 'kl':
+        nonzero = p_i > 1e-10
+        neighbours_proba[nonzero, :] /= p_i[nonzero, np.newaxis]
```

Comment:
IIRC, fancy indexing created a copy up until recent versions of numpy, so this assignment would be void:

```python
import numpy as np
a = np.array([0., 1.])
nnz = a != 0
print (a)  # [0., 1.]
a[nnz] += 1
print (a)  # can yield [0., 1.] in older numpy versions
```
It seems to be OK for 1.8 though (I don't have immediate access to lower versions right now and can't recall to which sklearn still supports, so don't change without making sure it's necessary)

In general, as Vlad says lower down, fancy indexing can slow things down. So check whether it is fully necessary here. It may be.